File: SSI/practica_1/ruby/lib/grc4.rb
require 'gtk2'
require_relative 'rc4'

class Rc4ui < Gtk::Window
  def initialize
    super
    @rc4 = Rc4.new "2 5"
    #Titulo de la aplicacion
    set_title "Algoritmo Rc4"
    set_icon Gdk::Pixbuf.new("image/logo.png")
    #Terminar aplicacion al cerrar ventana
    signal_connect "destroy" do
      quit
    end
    #Inicializar contenido de la ventana
    components
    #Establecer tamaño de la ventana y posicion de la ventana
    set_default_size 400,200
    set_window_position Gtk::Window::POS_CENTER
    #Mostrar ventana principal
    show_all
  end
  
  def components
    #Crear un fix donde se colocaran el resto de los widgets
    fixed = Gtk::Fixed.new
    add fixed
    #Boton salir
    btn_salir = Gtk::Button.new Gtk::Stock::QUIT
    btn_salir.signal_connect "clicked" do
      quit
    end
    #Crear los botones cifrar y about
    btn_cifrar = Gtk::Button.new "Cifrar"
    btn_cifrar.set_tooltip_text "Cifrar mensaje"
    btn_cifrar.signal_connect "clicked" do
      on_cifrar_clicked
    end
    btn_about = Gtk::Button.new Gtk::Stock::INFO
    btn_about.set_tooltip_text "Acerca del autor"
    btn_about.signal_connect "clicked" do
      on_about_clicked
    end
    #Introducir botones en el fixed
    fixed.put btn_cifrar, 100, 140
    fixed.put btn_about, 164, 140
    fixed.put btn_salir, 290, 140
    #Introducir etiquetas de los campos de texto
    lbl_seed = Gtk::Label.new "Semilla"
    lbl_message = Gtk::Label.new "Mensaje"
    lbl_result = Gtk::Label.new "Resultado"
    #Introducir etiquetas en el fixed
    fixed.put lbl_seed, 20,30
    fixed.put lbl_message, 20,60
    fixed.put lbl_result, 20,90
    #Crear los inputs para introducir la semilla, el mensaje y el resultado
    @in_seed = Gtk::Entry.new
    @in_seed.set_size_request 250,26
    @in_seed.set_text @rc4.seed.join(" ")
    @in_message = Gtk::Entry.new
    @in_message.set_size_request 250,26
    @in_message.set_text "1 34"
    @in_result = Gtk::Entry.new
    @in_result.set_size_request 250,26
    #Introducir los inputs en el fixed
    fixed.put @in_seed,120,25
    fixed.put @in_message,120,55
    fixed.put @in_result,120,85
    
  end
  
  def on_about_clicked
    about = Gtk::AboutDialog.new
    about.set_program_name "Algotirmo Rc4"
    about.set_version "0.1"
    about.set_copyright "(c) Victor Mora"
    about.run
    about.destroy
  end
  
  def on_cifrar_clicked
    #Todo esto hay que meterlo en un Thread!
    seed = @in_seed.text
    message = @in_message.text
    @rc4.seed= seed
    @rc4.code(message)
    @in_result.set_text @rc4.to_s
  end
  
  def quit
    Gtk.main_quit
    puts "App terminada"
  end
end

if __FILE__ == $0
  Gtk.init
  window = Rc4ui.new
  Gtk.main
end

File: SSI/practica_1/cpp/lib/grc4.cc
#include "grc4.h"
#include <iostream>

Grc4::Grc4(): btn_quit(Gtk::Stock::QUIT),
			  btn_info(Gtk::Stock::INFO),
			  lbl_seed("Semilla"),
			  lbl_message("Mensaje"),
			  lbl_result("Resultado"),
			  about(),
			  rc4()
{
  // Sets the border width of the window.
  set_default_size(400,200);
  set_title("Algoritmo Rc4");
  //set_window_position(Gtk::Window::POS_CENTER);
  set_icon(Gdk::Pixbuf::create_from_file("image/logo.png"));
  set_position(Gtk::WIN_POS_CENTER);
  components();
  show_all_children();
}

void Grc4::components() {
	//Boton para cifrar
	btn_code.add_label("Cifrar");
	btn_code.set_tooltip_text("Cifrar mensaje");
	btn_code.signal_clicked().connect(sigc::mem_fun(*this,&Grc4::on_btn_code_clicked));
	//Boton salir
	btn_quit.signal_clicked().connect(sigc::mem_fun(*this,&Grc4::on_btn_quit_clicked));
	//Boton boton about
	btn_info.signal_clicked().connect(sigc::mem_fun(*this,&Grc4::on_btn_info_clicked));
	//Inputs
	in_seed.set_size_request(250,26);
	in_message.set_size_request(250,26);
	in_result.set_size_request(250,26);
	//Introducir botones en el fixed
	fixed.put(btn_code,100,140);
	fixed.put(btn_info,164,140);
	fixed.put(btn_quit,290,140);
	//Introducir etiquetas en el fixed
	fixed.put(lbl_seed,20,30);
	fixed.put(lbl_message,20,60);
	fixed.put(lbl_result,20,90);
	//Introducir inputs en el fixed
	fixed.put(in_seed,120,25);
	fixed.put(in_message,120,55);
	fixed.put(in_result,120,85);
	//Introducir el fixed en la ventana
	add(fixed);
}

void Grc4::on_btn_code_clicked()
{
	rc4.set_seed(in_seed.get_text());
	rc4.code(in_message.get_text());
	std::cout << rc4.to_s() << std::endl;
	in_result.set_text(rc4.to_s());
}

void Grc4::on_btn_info_clicked()
{
	about.set_size_request(50,50);
	about.set_program_name("Algoritmo Rc4");
	about.set_version("0.1");
	about.set_copyright("(c) Victor Mora");
	about.run();
	about.~AboutDialog();
}

void Grc4::on_btn_quit_clicked()
{
	Gtk::Main::quit();
	std::cout << "App terminada" << std::endl;
}


